### PR TITLE
Add recipe for ptxcompiler

### PR DESCRIPTION
**Pull request by prashantkhoje**
Recipe to build ptxcompiler

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/ptxcompiler/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  missing_dso_whitelist:
+    - '*libc.so*'        # [linux]
+    - '*libstdc++.so*'   # [linux]
+    - '*libpthread.so*'  # [linux]
+    - '*libm.so*'        # [linux]
+    - '*libgcc_s.so*'    # [linux]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:31]: I think you need to depend on the CUDA compiler.

Comment (hmaarrfk) [reply recipes/ptxcompiler/meta.yaml:31]: See the pytorch or tensorflow recipes

File: recipes/ptxcompiler/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+source:
+  - url: https://github.com/rapidsai/ptxcompiler/archive/refs/tags/{{ version }}.tar.gz
+    sha256: c7021a0946478404c4055a612fd725af95d53ec23eb3dd8d8bc0b5ac55eb44d0
+    folder: ptxcompiler
+
+build:
+  number: 0
+  skip: true  # [(not linux) or (cuda_compiler_version !=  "11.2")]
+  script_env:
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:16]: Also remove these. They will be set for you by cuda compilers and cudnn

File: recipes/ptxcompiler/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
+{% set name = "ptxcompiler" %}
+{% set version = "0.3.0" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  - url: https://github.com/rapidsai/ptxcompiler/archive/refs/tags/{{ version }}.tar.gz
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:9]: If you remove ethe `- ` and the `fn` you want have to use `cd` below.

File: recipes/ptxcompiler/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  number: 0
+  skip: true  # [(not linux) or (cuda_compiler_version !=  "11.2")]
+  script_env:
+    - CPATH=/usr/local/cuda/include
+    - LIBRARY_PATH=/usr/local/cuda/lib64
+  script:
+    - cd ptxcompiler
+    - {{ PYTHON }} -m pip install . -vv
+  missing_dso_whitelist:
+    - '*libc.so*'        # [linux]
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:23]: I'm unconvinced that you absolutely need these. Typically this indicates a missing dependency

File: recipes/ptxcompiler/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    - '*libstdc++.so*'   # [linux]
+    - '*libpthread.so*'  # [linux]
+    - '*libm.so*'        # [linux]
+    - '*libgcc_s.so*'    # [linux]
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}  # [cuda_compiler_version != "None"]
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:32]: Maybe add a dependency on the c compiler.

Comment (hmaarrfk) [reply recipes/ptxcompiler/meta.yaml:32]: In addition to the cpp compiler

File: recipes/ptxcompiler/conda_build_config.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+docker_image:                                           # [os.environ.get("BUILD_PLATFORM", "").startswith("linux")]
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/conda_build_config.yaml:1]: Is this file really necessary now? Can it not wait for the migration?

File: recipes/ptxcompiler/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  version: {{ version }}
+
+source:
+  url: https://github.com/rapidsai/ptxcompiler/archive/refs/tags/{{ version }}.tar.gz
+  sha256: c7021a0946478404c4055a612fd725af95d53ec23eb3dd8d8bc0b5ac55eb44d0
+  fn: ptxcompiler.tar.gz
+
+build:
+  number: 0
+  skip: true  # [(not linux) or (cuda_compiler_version !=  "None")]
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:15]:
```suggestion
  skip: true  # [(not linux) or (cuda_compiler_version ==  "None")]
```

Should this be the condition?

Comment (prashantkhoje) [reply recipes/ptxcompiler/meta.yaml]: Thank you for the review.
Without env variables build fails to get nvPTXCompiler.h.
The ptxcompiler requires cuda 11.1 or higher. How can i skip build for lower versions?

**Comment by hmaarrfk** [comment]
Does it only support 11.1 and higher?

**Comment by prashantkhoje** [comment]
> Does it only support 11.1 and higher?
Yes. The ptxcompiler was first introduced in 11.1 - 
https://github.com/rapidsai/ptxcompiler/blob/95cafed62e267fae25f8e1c34ace00519bf82838/ptxcompiler/tests/test_lib.py#L51

**Comment by hmaarrfk** [comment]
Can you skip those as well? In the migration process they will show up as errors

File: recipes/ptxcompiler/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  version: {{ version }}
+
+source:
+  url: https://github.com/rapidsai/ptxcompiler/archive/refs/tags/{{ version }}.tar.gz
+  sha256: c7021a0946478404c4055a612fd725af95d53ec23eb3dd8d8bc0b5ac55eb44d0
+  fn: ptxcompiler.tar.gz
+
+build:
+  number: 0
+  skip: true  # [(not linux) or (cuda_compiler_version == "None")]
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:15]:
```suggestion
  skip: true  # [not linux]
  skip: true  # [cuda_compiler_version == "None"]
  skip: true  # [cuda_compiler_version == "10.2"]
  skip: true  # [cuda_compiler_version == "11.0"]
```

I get confused with too many statements in one line.

I think you can also use syntax like
```
cuda_compiler_version in ["None", "10.2", "11.0"]
```

File: recipes/ptxcompiler/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  fn: ptxcompiler.tar.gz
+
+build:
+  number: 0
+  skip: true  # [not linux]
+  skip: true  # [cuda_compiler_version == "None"]
+  skip: true  # [cuda_compiler_version == "10.2"]
+  skip: true  # [cuda_compiler_version == "11.0"]
+  script_env:
+    - CPATH=/usr/local/cuda/include
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:20]: Have you tried removing these definitions?

Comment (prashantkhoje) [reply recipes/ptxcompiler/meta.yaml]: Yes. Build fails with: [cannot find -lnvptxcompiler_static](https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/470564/logs/61)
I also tried using CUDA_HOME (as in https://github.com/conda-forge/rmm-feedstock/blob/main/recipe/meta.yaml) instead of CPATH and LIBRARY_PATH.
I'll restore these env vars now.

File: recipes/ptxcompiler/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  fn: ptxcompiler.tar.gz
+
+build:
+  number: 0
+  skip: true  # [not linux]
+  skip: true  # [cuda_compiler_version == "None"]
+  skip: true  # [cuda_compiler_version == "10.2"]
+  skip: true  # [cuda_compiler_version == "11.0"]
+  script_env: 
+    - CUDA_HOME  # [ppc64le or aarch64]
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:20]:
```suggestion
```

I'm really confused why `CUDA_HOME` is being added at this stage. As far as conda-forge is concerned, there are no cuda builds with `ppc64le` or `aarch64` therefore, they would all be skipped.

File: recipes/ptxcompiler/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    - LIBRARY_PATH=/usr/local/cuda/lib64
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:21]: It seems that this patch:
https://github.com/rapidsai/ptxcompiler/commit/5766ac29aab9ddba3a160d45d5d86cfbd3c9a647
was recently submitted, to help you avoid this export.

Can you cut a new release and try building without this customization.

File: recipes/ptxcompiler/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+{% set version = "0.3.0" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://github.com/rapidsai/ptxcompiler/archive/refs/tags/{{ version }}.tar.gz
+  sha256: c7021a0946478404c4055a612fd725af95d53ec23eb3dd8d8bc0b5ac55eb44d0
+  fn: ptxcompiler.tar.gz
```
Comment (hmaarrfk) [inline recipes/ptxcompiler/meta.yaml:11]:
```suggestion
```

**Comment by hmaarrfk** [comment]
This looks like it is getting there. Hopefully recent additions to your library will make it more broadly compatibile with conda-forge. Please ping when you've cut a new release and or patched this one.

**Comment by hmaarrfk** [comment]
Looks great. Sorry about all the nits. CUDA has been tricky and we have been working hard to keep things organized.

We should likely do a better job documenting all this in a central location.

**Review by hmaarrfk — APPROVED** [review]

**Comment by prashantkhoje** [comment]
> Looks great. Sorry about all the nits. CUDA has been tricky and we have been working hard to keep things organized.
> 
> We should likely do a better job documenting all this in a central location.

Thank you for the review and inputs!

**Comment by jakirkham** [comment]
cc @ajschmidt8 @gmarkall @quasiben (for viz)

**Comment by jakirkham** [comment]
Should add the way this package works is we need to build with the latest CUDA version and then it will support any earlier versions of CUDA. So this is the opposite of how conda-forge is doing things with CUDA Enhanced Compatibility. This may require some more work in the feedstock to get this working correctly. Please see [this recipe]( https://github.com/rapidsai/ptxcompiler/blob/eed289a36a68e252dc7bd2888b947c801756eec7/conda/recipes/ptxcompiler/meta.yaml#L20-L33 ) for details

**Comment by quasiben** [comment]
cc @bdice who was also interested in better packaging support for ptxcompiler
